hare3/protocol.go

```diff
@@ -0,0 +1,451 @@
+}
+
+func toHash(proposals []types.ProposalID) types.Hash32 {
+	return types.CalcProposalHash32Presorted(proposals, nil)
+}
+
+type messageKey struct {
+	IterRound
+	Sender types.NodeID
+}
+
+type input struct {
+	*Message
+	atxgrade  grade
+	malicious bool
+	msgHash   types.Hash32
+}
+
+func (i *input) MarshalLogObject(encoder zapcore.ObjectEncoder) error {
+	if i.Message != nil {
+		i.Message.MarshalLogObject(encoder)
+	}
+	encoder.AddUint8("atxgrade", uint8(i.atxgrade))
+	encoder.AddBool("malicious", i.malicious)
+	encoder.AddString("hash", i.msgHash.ShortString())
+	return nil
+}
+
+type output struct {
+	coin       *bool              // set based on preround messages right after preround completes in 0 iter
+	result     []types.ProposalID // set based on notify messages at the start of next iter
+	terminated bool               // protocol participates in one more iteration after outputing result
+	message    *Message
+}
+
+func (o *output) MarshalLogObject(encoder zapcore.ObjectEncoder) error {
+	encoder.AddBool("terminated", o.terminated)
+	if o.coin != nil {
+		encoder.AddBool("coin", *o.coin)
+	}
+	if o.result != nil {
+		encoder.AddArray("result", zapcore.ArrayMarshalerFunc(func(encoder log.ArrayEncoder) error {
+			for _, id := range o.result {
+				encoder.AppendString(types.Hash20(id).ShortString())
+			}
+			return nil
+		}))
+	}
+	if o.message != nil {
+		encoder.AddObject("msg", o.message)
+	}
+	return nil
+}
+
+func newProtocol(initial []types.ProposalID, threshold uint16) *protocol {
+	return &protocol{
+		initial:        initial,
+		validProposals: map[types.Hash32][]types.ProposalID{},
+		gradedGossip:   gradedGossip{state: map[messageKey]*input{}},
+		gradecast:      gradecast{state: map[messageKey]*gradecasted{}},
+		thresholdGossip: thresholdGossip{
+			threshold: threshold,
+			state:     map[messageKey]*votes{},
+		},
+	}
+}
+
+type protocol struct {
+	IterRound
+	coinout         bool
+	coin            *types.VrfSignature // smallest vrf from preround messages. not a part of paper
+	initial         []types.ProposalID  // Si
+	result          *types.Hash32       // set after Round 6. Case 1
+	locked          *types.Hash32       // Li
+	hardLocked      bool
+	validProposals  map[types.Hash32][]types.ProposalID // Ti
+	validValues     [grade5 + 1][]types.ProposalID      // Vi
+	gradedGossip    gradedGossip                        // Protocol 1. graded-gossip. page 10.
+	gradecast       gradecast                           // Protocol 2. gradecast. page 13.
+	thresholdGossip thresholdGossip                     // Protocol 3. thresh-gossip. Page 15.
+}
+
+func (p *protocol) onInput(msg *input) (bool, *types.HareProof) {
+	gossip, equivocation := p.gradedGossip.receive(msg)
+	if !gossip {
+		return false, equivocation
+	}
+	// gradecast and thresholdGossip should never be called with non-equivocating duplicates
+	if msg.Round == propose {
+		p.gradecast.receive(p.IterRound, msg)
+	} else {
+		p.thresholdGossip.receive(msg)
+	}
```

Comment:
> i had imagined that upon invocation of gradecast and thresholdGossip, each sub protocol will just take the state of gradedGossip as input and output the result.

so in practice it can be reduced to single map instead of 3 different? should be possible, will try that

Reply:
thanks, much more compact now